src/frontend/hooks/useSyncState.ts

```diff
@@ -29,68 +38,132 @@ function identity(state: SyncState | undefined) {
+ */
+export function useSyncProgress() {
+  const {subscribe, getProgressSnapshot} = useSyncStore();
+  return useSyncExternalStore(subscribe, getProgressSnapshot);
+}
+
+class SyncStore {
+  #project: MapeoProjectApi;
+
+  #listeners = new Set<() => void>();
+  #isSubscribedInternal = false;
+  #error: Error | null = null;
+  #state: SyncState | null = null;
+
+  /**
+   * Represents maximum value of `#state.data.want + #state.data.wanted` while data syncing is enabled.
+   * Resets to null when data syncing goes from enabled to disabled.
+   */
+  #maxDataSyncCount: number | null = null;
+
+  constructor(project: MapeoProjectApi) {
+    this.#project = project;
+  }
+
+  subscribe = (listener: () => void) => {
+    this.#listeners.add(listener);
+    if (!this.#isSubscribedInternal) this.#startSubscription();
+    return () => {
+      this.#listeners.delete(listener);
+      if (this.#listeners.size === 0) this.#stopSubscription();
+    };
+  };
+
+  getStateSnapshot = () => {
+    if (this.#error) throw this.#error;
+    return this.#state;
+  };
+
+  getProgressSnapshot = () => {
+    if (this.#maxDataSyncCount === null || this.#state === null) {
+      return null;
+    }
+
+    if (this.#maxDataSyncCount === 0) {
+      return 1;
+    }
+
+    const currentCount = this.#state.data.want + this.#state.data.wanted;
+
+    const ratio =
+      (this.#maxDataSyncCount - currentCount) / this.#maxDataSyncCount;
+
+    if (ratio <= 0) return 0;
+    if (ratio >= 1) return 1;
 
-  function onSyncState(newState: SyncState) {
-    state = newState;
-    error = undefined;
-    listeners.forEach(listener => listener());
+    return clamp(ratio, 0.01, 0.99);
+  };
+
+  #notifyListeners() {
+    for (const listener of this.#listeners) {
+      listener();
+    }
   }
 
-  function subscribeInternal() {
-    project.$sync.on('sync-state', onSyncState);
-    isSubscribedInternal = true;
-    project.$sync
+  #onSyncState = (state: SyncState) => {
+    console.log({from: 'listener', ...state});
+    // Indicates whether data syncing went from enabled to disabled
+    const isDataSyncStopped = this.#state?.data.syncing && !state.data.syncing;
+
+    if (isDataSyncStopped) {
+      this.#maxDataSyncCount = null;
+    } else {
+      const newSyncCount = state.data.want + state.data.wanted;
+
+      this.#maxDataSyncCount =
+        this.#maxDataSyncCount === null
+          ? newSyncCount
+          : Math.max(this.#maxDataSyncCount, newSyncCount);
+    }
+
+    this.#state = state;
+    this.#error = null;
+    this.#notifyListeners();
+  };
+
+  #startSubscription = () => {
+    this.#project.$sync.on('sync-state', this.#onSyncState);
+    this.#isSubscribedInternal = true;
+    this.#project.$sync
       .getState()
-      .then(onSyncState)
+      .then(this.#onSyncState)
       .catch(e => {
-        error = e;
-        listeners.forEach(listener => listener());
+        this.#error = e;
+        this.#notifyListeners();
```

Comment:
Should we do anything special if we've unsubscribed before this promise resolves? I imagine the following scenario:

1. `#startSubscription` is called, kicking off `this.#project.$sync.getState()`.
2. `#stopSubscription` is called.
3. The `getState()` promise resolves.

Reply:
i _think_ it's okay because resolving will eventually call the subscribed listeners, but if `#stopSubscription` is called it's because there are no more listeners, so it should result in a no-op.

that's my initial thinking, anyways